### PR TITLE
Automatically revive hopeless kids on reconfigure and after a timeout

### PR DESCRIPTION
**Pull request by eduard-bagdasaryan**
Squid declares kids with "repeated, frequent failures" as hopeless.
Hopeless kids were not automatically restarted. In the absence of
automated recovery, admins were forced to restart the whole Squid
instance (after fixing the underlying problem that led to those kid
failures). Squid instance restarts hurt users.

In many cases, the underlying kid-killing problem is temporary, and
Squid can eventually fully recover without any admin involvement.

Squid now automatically restarts a hopeless kid after a configurable
downtime (a new hopeless_kid_revival_delay directive with a 60 minute
default value).

We also now restart all hopeless kids upon receiving a reconfiguration
signal.

Also: do not send signals to non-running kids.

### CODE REVIEW

File: src/main.cc [hunk 1]
```diff
@@ -1951,55 +2046,25 @@ watch_child(char *argv[])
...
-                exit(EXIT_FAILURE);
-            }
-
-            if (TheKids.allHopeless()) {
-                syslog(LOG_ALERT, "Exiting due to repeated, frequent failures");
-                exit(EXIT_FAILURE);
-            }
-
-            exit(EXIT_SUCCESS);
+            MasterExit();
```
Comment (rousskov) [inline src/main.cc]: Please `s/MasterExit/masterExit/` for consistency sake. We should eventually capitalize all of the static `master*()` functions or, better, move them into a new class, but either is outside this project scope.

File: src/main.cc [hunk 2]
```diff
@@ -1817,6 +1887,22 @@ GoIntoBackground()
 }
 #endif /* !_SQUID_WINDOWS_ */
 
+
+static void MasterExit()
```
Comment (rousskov) [inline src/main.cc]: If there is no particular reason to keep MasterExit() available on Windows, then I would lower `#endif /* !_SQUID_WINDOWS_ */` to cover this function as well. Doing so will be consistent with many other local static functions that are only called from `watch_child()`. It might also help prevent errors or warnings when this signal-using code gets compiled on Windows.

File: src/cf.data.pre [hunk 3]
```diff
@@ -473,6 +473,25 @@ DOC_START
...
+TYPE: time_t
+LOC: Config.hopelessKidRevivalDelay
+DEFAULT: 60 minutes
+DOC_START
+	Normally, when a kid process dies, Squid immediately restarts the
+	kid. A kid experiencing frequent deaths is marked as "hopeless" for
+	the duration specified by this directive. Hopeless kids are not
+	automatically restarted.
+
+	To always restart kids, even the ones that are experiencing frequent
```
Comment (rousskov) [inline src/cf.data.pre]: This paragraph is out of sync with the code. Consider this replacement:

```
Currently, zero values are not supported because they result in
misconfigured SMP Squid instances running forever, endlessly
restarting each dying kid. To effectively disable hopeless kids
revival, set the delay to a huge value (e.g., 10 years).
```

File: src/main.cc [hunk 4]
```diff
@@ -1830,6 +1916,12 @@ watch_child(char *argv[])
 
     int nullfd;
 
+    // XXX: zero values are not supported because they result in
```
Comment (rousskov) [inline src/main.cc]: I suggest replacing this `XXX` label with `TODO`. A slight restriction on configuration values does not represent a bug IMO. Eventually, somebody may find a way to support them and address this TODO without breaking any existing configurations (because we prohibit zero values).

File: src/main.cc [hunk 5]
```diff
@@ -1830,6 +1916,12 @@ watch_child(char *argv[])
...
+    // misconfigured SMP Squid instances running forever, endlessly
+    // restarting each dying kid.
+    if (Config.hopelessKidRevivalDelay == 0)
```
Comment (rousskov) [inline src/main.cc]: Please use `<= 0` to better match the error message (even if some other code is already testing that configured time_t values cannot be negative).

**Comment by rousskov** [comment]
I took the liberty to move the commit title into the PR title. I think that is what Github expects when merging the PR, but please correct me if I guessed wrong.

Also, the internal issue references belong to branch names, not commit messages because that developer-specific housekeeping info is not related to the Squid Project.

**Comment by rousskov** [comment]
@squid-cache/commiters, if this PR is approved, please squash before merging. Eventually, we should have a policy about that (which should, in part, prevent untested squash commits, but that requires more tooling than we currently have).

File: src/cf.data.pre (hunk 3 above)
Comment (eduard-bagdasaryan) [inline src/cf.data.pre]: I fixed this just after PR creation but pushed with a delay...

File: src/main.cc (hunk 5 above)
Comment (eduard-bagdasaryan) [inline src/main.cc]: Done.

File: src/main.cc (hunk 4 above)
Comment (eduard-bagdasaryan) [inline src/main.cc]: Done.

File: src/main.cc (hunk 2 above)
Comment (eduard-bagdasaryan) [inline src/main.cc]: Done.

File: src/cf.data.pre [hunk 6]
```diff
@@ -473,6 +473,26 @@ DOC_START
 	CAP_IPC_LOCK capability, or equivalent.
 DOC_END
 
+NAME: hopeless_kid_revival_delay
+COMMENT: time-units
+TYPE: time_t
+LOC: Config.hopelessKidRevivalDelay
+DEFAULT: 60 minutes
```
Comment (yadij) [inline src/cf.data.pre]: parseTimeUnits() does not accept plurals and the definition of T_MINUTE_STR is "minute". You may want to use "1 hour" though instead of "60 minute"

File: src/cf.data.pre [hunk 7]
```diff
@@ -473,6 +473,26 @@ DOC_START
...
+DOC_START
+	Normally, when a kid process dies, Squid immediately restarts the
+	kid. A kid experiencing frequent deaths is marked as "hopeless" for
+	the duration specified by this directive. Hopeless kids are not
+	automatically restarted.
+
+	Currently, zero values are not supported because they result in
+	misconfigured SMP Squid instances running forever, endlessly
+	restarting each dying kid. To effectively disable hopeless kids
+	revival, set the delay to a huge value (e.g., 10 years).
```
Comment (yadij) [inline src/cf.data.pre]: "1 year" should be long enough and also avoids the issue with plurals.

File: src/ipc/Kid.cc [hunk 8]
```diff
@@ -11,6 +11,8 @@
 #include "squid.h"
 #include "globals.h"
 #include "ipc/Kid.h"
+#include "SquidConfig.h"
+#include "SquidTime.h"
```
Comment (yadij) [inline src/ipc/Kid.cc]: SquidConf.h already includes SquidTime.h please remove this unnecessary line.

File: src/ipc/Kids.cc [hunk 9]
```diff
@@ -127,4 +156,3 @@ size_t Kids::count() const
 {
     return storage.size();
 }
-
```
Comment (yadij) [inline src/ipc/Kids.cc]: unnecessary change

File: src/ipc/Kid.cc [hunk 10]
```diff
@@ -23,6 +25,7 @@ Kid::Kid():
     badFailures(0),
     pid(-1),
     startTime(0),
+    stopTime(0),
```
Comment (yadij) [inline src/ipc/Kid.cc]: Please use the C++11 initialization style for new members. That will allow removal of the code duplication here and the next chunk.

File: src/ipc/Kid.cc [hunk 11]
```diff
@@ -57,15 +62,39 @@ Kid::stop(PidStatus const theExitStatus)
...
     else
         badFailures = 0; // the failures are not "frequent" [any more]
 
-    status = theExitStatus;
+    reportStopped(); // after all state changes
+}
+
+/// describes a recently stopped kid
+void
+Kid::reportStopped() const
```
Comment (yadij) [inline src/ipc/Kid.cc]: Is use of debugs() possible here? either in place of or as well as the syslog calls.

File: src/main.cc [hunk 12]
```diff
@@ -1774,32 +1789,87 @@ mainStartScript(const char *prog)
...
+    reconfiguring = 0;
+}
+
+/// Reacts to the kid revival alarm.
+static void
+masterReviveKids()
+{
+    if (AvoidSignalAction("kids revival", do_revive_kids))
+        return;
+    debugs(1, 2, "woke up after ~" << Config.hopelessKidRevivalDelay << "s");
```
Comment (yadij) [inline src/main.cc]: Please use DBG_IMPORTANT instead of '1'

File: src/main.cc [hunk 13]
```diff
@@ -1774,32 +1789,87 @@ mainStartScript(const char *prog)
...
+/// Maintains the following invariant: An alarm should be scheduled when and
+/// only when there are hopeless kid(s) that cannot be immediately revived.
+static void
+masterMaintainKidRevivalSchedule()
+{
+    const auto nextCheckDelay = TheKids.forgetOldFailures();
+    assert(nextCheckDelay >= 0);
+    (void)alarm(static_cast<unsigned int>(nextCheckDelay)); // resets or cancels
+    if (nextCheckDelay)
+        debugs(1, 2, "will recheck hopeless kids in " << nextCheckDelay << " seconds");
```
Comment (yadij) [inline src/main.cc]: Please use DBG_IMPORTANT instead of '1'.

File: src/main.cc [hunk 14]
```diff
@@ -1774,32 +1789,87 @@ mainStartScript(const char *prog)
...
+/// Initiates reconfiguration sequence. See also: masterReconfigureFinish().
 static void
-masterCheckAndBroadcastSignals()
+masterReconfigureStart()
 {
-    // if (do_reconfigure)
-    //     TODO: hot-reconfiguration of the number of kids and PID file location
+    if (AvoidSignalAction("reconfiguration", do_reconfigure))
+        return;
+    debugs(1, 2, "received reconfiguration command");
```
Comment (yadij) [inline src/main.cc]: Please use DBG_IMPORTANT instead of '1'.

File: src/cf.data.pre (hunk 6 above)
Comment (eduard-bagdasaryan) [inline src/cf.data.pre]: Though you a right that current squid time parsing code is far from ideal and allows such oddities as '60 minute', the current squid default configuration documentation contains lots of such examples, e.g., 'peer_connect_timeout 30 seconds', read_timeout 15 minutes', 'positive_dns_ttl 6 hours' and so on. The current 60 minutes Config.hopelessKidRevivalDelay does not make worse the existing order of things. Probably we need to improve the existing parsing code, but this is out of scope.
Leaving as is for now.

File: src/cf.data.pre (hunk 7 above)
Comment (eduard-bagdasaryan) [inline src/cf.data.pre]: Leaving as is for now for a reason explained [elsewhere](https://github.com/squid-cache/squid/pull/32/files/cdfd9f57d523efe906d4e75773ce23f8a77a62ee#r130610921).

File: src/ipc/Kid.cc (hunk 11 above)
Comment (eduard-bagdasaryan) [inline src/ipc/Kid.cc]: I am not against making this 'debugs' or adding some 'debugs', however there may be many other 'syslog' calls in master code needing similar change? I would do all that changes in a separate issue. IIRC, we were also thinking about using debugs everywhere and redirecting messages to syslog by _db_set_syslog() method.

File: src/ipc/Kid.cc (hunk 8 above)
Comment (eduard-bagdasaryan) [inline src/ipc/Kid.cc]: Done.

File: src/ipc/Kids.cc (hunk 9 above)
Comment (eduard-bagdasaryan) [inline src/ipc/Kids.cc]: Done.

File: src/ipc/Kid.cc (hunk 10 above)
Comment (eduard-bagdasaryan) [inline src/ipc/Kid.cc]: Done.

File: src/cf.data.pre (hunk 6 above)
Comment (rousskov) [inline src/cf.data.pre]: Since `60 minute` is grammatically incorrect, the current parsing code supports `60 minutes`, and many existing defaults already use `minutes`, I think `60 minutes` is acceptable. However, please use `1 hour` instead to avoid debating a very minor issue.

File: src/cf.data.pre (hunk 7 above)
Comment (rousskov) [inline src/cf.data.pre]: Since `1 year` does not sound `huge`, the current parsing code supports `10 years` (I presume), and many existing defaults already use plural units, I think `10 years` is acceptable. However, please use `1 year` instead to avoid debating a minor issue.

File: src/ipc/Kid.cc (hunk 11 above)
Comment (rousskov) [inline src/ipc/Kid.cc]: This is _moved_ code. Any changes can be considered out of scope.

Please do not add debugs(). We do not want to duplicate the reporting code.

Please do not replace syslog() with debugs(). Many admins rely on these messages going to syslog.

This problem will disappear when we start using debugs() instead of syslog() in the daemon process _and_ copy level-0/1 messages to syslog(). That change is being discussed on squid-dev and is outside this project scope.

File: src/cf.data.pre (hunk 7 above)
Comment (eduard-bagdasaryan) [inline src/cf.data.pre]: Done.

File: src/cf.data.pre (hunk 6 above)
Comment (eduard-bagdasaryan) [inline src/cf.data.pre]: Done.

File: src/ipc/Kid.cc (hunk 8 above)
Comment (rousskov) [inline src/ipc/Kid.cc]: I am not against this change, but, for the record, we should not expect humans to minimize `#include` lists. Including file X in file Y when X declares something used in Y is fine, regardless of whether Y also includes some other file Z which also includes X.

IMHO, in general, the list of includes in Z is not a part of Z's public API -- it provides no guarantees and may change at any time. For example,  Z might include X to declare some private Z class member; changes in such private declarations ought to have no effect on Z users.

**Comment by rousskov** [comment]
@yadij , thank you for approving this PR. I am taking the time to test and document PR rebasing/squashing procedure using this PR as a Guinea pig...
